htlcswitch/quiescer.go

```diff
@@ -0,0 +1,233 @@
+	ErrNoQuiescenceInitiator = fmt.Errorf(
+		"indeterminate quiescence initiator: channel is not quiescent",
+	)
+
+	// ErrPendingRemoteUpdates indicates that we have received an Stfu while
+	// the remote party has issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingRemoteUpdates = fmt.Errorf(
+		"stfu received with pending remote updates",
+	)
+
+	// ErrPendingLocalUpdates indicates that we are attempting to send an
+	// Stfu while we have issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingLocalUpdates = fmt.Errorf(
+		"stfu send attempted with pending local updates",
+	)
+)
+
+type quiescerCfg struct {
+	// chanID marks what channel we are managing the state machine for. This
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// channelInitiator indicates which ChannelParty originally opened the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	channelInitiator lntypes.ChannelParty
+
+	// numPendingUpdates is a function that returns the number of pending
+	// originated by the party in the first argument that have yet to be
+	// committed to the commitment transaction held by the party in the
+	// second argument.
+	numPendingUpdates func(lntypes.ChannelParty,
+		lntypes.ChannelParty) uint64
+
+	// sendMsg is a function that can be used to send an Stfu message over
+	// the wire.
+	sendMsg func(lnwire.Stfu) error
+}
+
+// quiescer is a state machine that tracks progression through the quiescence
+// protocol.
+type quiescer struct {
+	cfg quiescerCfg
+
+	// localInit indicates whether our path through this state machine was
+	// initiated by our node. This can be true or false independently of
+	// remoteInit.
+	localInit bool
+
+	// remoteInit indicates whether we received Stfu from our peer where the
+	// message indicated that the remote node believes it was the initiator.
+	// This can be true or false independently of localInit.
+	remoteInit bool
+
+	// sent tracks whether or not we have emitted Stfu for sending.
+	sent bool
+
+	// received tracks whether or not we have received Stfu from our peer.
+	received bool
+}
+
+// newQuiescer creates a new quiescer for the given channel.
+func newQuiescer(cfg quiescerCfg) quiescer {
+	return quiescer{
+		cfg: cfg,
+	}
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescer) recvStfu(msg lnwire.Stfu) error {
+	// At the time of this writing, this check that we have already received
+	// an Stfu is not strictly necessary, according to the specification.
+	// However, it is fishy if we do and it is unclear how we should handle
+	// such a case so we will err on the side of caution.
+	if q.received {
+		return fmt.Errorf("%w for channel %v", ErrStfuAlreadyRcvd,
```

Comment:
Re the comment above, this would just be an invalid state transition based on the incomin event and current state (vs an assertion on some attribute).

Reply:
Yeah either way it's an error, but I suppose with the FSM construction it would be a little clearer. I can redesign this state machine to work this way if you think it's worth it. I do think that it is clear and simple enough as is though, though the enum states are probably better.